cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/TopicReplicationFactorPlan.java

```diff
@@ -0,0 +1,111 @@
+      if (regex == null || regex.isEmpty()) {
+        throw new IllegalArgumentException("Missing topics of the replication factor update for the plan.");
+      }
+    }
+    _topicRegexWithRFUpdate = topicRegexWithRFUpdate;
+  }
+
+  @Override
+  public byte planVersion() {
+    return PLAN_VERSION;
+  }
+
+  public Map<Short, String> topicRegexWithRFUpdate() {
+    return _topicRegexWithRFUpdate;
+  }
+
+  @Override
+  public ByteBuffer toBuffer(int headerSize) {
+    byte numRFUpdateEntries = (byte) _topicRegexWithRFUpdate.size();
+    int requiredCapacityForRFEntries = 0;
+    for (Map.Entry<Short, String> entry : _topicRegexWithRFUpdate.entrySet()) {
+      requiredCapacityForRFEntries += (Short.BYTES /* replication factor */ + Integer.BYTES /* regex length */
+                                       + entry.getValue().getBytes(StandardCharsets.UTF_8).length /* regex */);
+    }
+    int contentSize = (Byte.BYTES /* plan version */
+                       + Long.BYTES /* timeMs */
+                       + Integer.BYTES /* broker id */
+                       + Byte.BYTES /* number of replication factor update entries */
+                       + requiredCapacityForRFEntries /* total capacity for all entries */);
+    ByteBuffer buffer = ByteBuffer.allocate(headerSize + Long.BYTES /* crc */ + contentSize);
+    buffer.position(headerSize + Long.BYTES);
+    buffer.put(planVersion());
+    buffer.putLong(timeMs());
+    buffer.putInt(brokerId());
+    buffer.put(numRFUpdateEntries);
+    for (Map.Entry<Short, String> entry : _topicRegexWithRFUpdate.entrySet()) {
+      buffer.putShort(entry.getKey());
+      byte[] regex = entry.getValue().getBytes(StandardCharsets.UTF_8);
+      buffer.putInt(regex.length);
+      buffer.put(regex);
+    }
+    putCrc(headerSize, buffer, contentSize);
+    return buffer;
+  }
+
+  /**
+   * Deserialize given byte buffer to an {@link TopicReplicationFactorPlan}.
+   *
+   * @param headerSize The header size of the buffer.
+   * @param buffer buffer to deserialize.
+   * @return The {@link TopicReplicationFactorPlan} corresponding to the deserialized buffer.
+   */
+  public static TopicReplicationFactorPlan fromBuffer(int headerSize, ByteBuffer buffer) throws UnknownVersionException {
+    verifyCrc(headerSize, buffer);
+    byte version = buffer.get();
+    if (version > PLAN_VERSION) {
+      throw new UnknownVersionException("Cannot deserialize the plan for version " + version + ". Current version: " + PLAN_VERSION);
+    }
+
+    long timeMs = buffer.getLong();
+    int brokerId = buffer.getInt();
+    byte numRFUpdateEntries = buffer.get();
+    Map<Short, String> topicRegexWithRFUpdate = new HashMap<>(numRFUpdateEntries);
+
+    for (int i = 0; i < numRFUpdateEntries; i++) {
+      short replicationFactor = buffer.getShort();
+      int regexLength = buffer.getInt();
+      String regex = new String(buffer.array(), buffer.arrayOffset() + buffer.position(), regexLength, StandardCharsets.UTF_8);
```

Comment:
Do you want to validate the regex here?

Reply:
The validity of the expression's syntax is verified in `MaintenanceEvent#topicPatternByReplicationFactor(Map<String, ?>)` -- i.e. upon creation of a `MaintenanceEvent` from `TopicReplicationFactorPlan`.

Reply:
That's assuming your code is going to be generating these events which it may not be.

Reply:
Yes, because the semantic interpretation of this String, called `regex`, is relevant to `MaintenanceEvent`.